4799 extension template documentation

f-osorio opened the pull request:
Fixes #4799 

Provides a new template for extensions using [Cookiecutter](https://cookiecutter.readthedocs.io/en/latest/index.html) instead of `Paster`. 

This only updates the documentation for the extension tutorial. The cookecutter [template is here](https://github.com/f-osorio/cookiecutter-ckan-extension) -- I wasn't sure where it should go since part of the reason for moving to cookiecutter seems to be to decouple it from CKAN core. The documentation only has a place holder to refer to the location until it's finalized.

smotornyuk commented:
As mentioned in the original issue, one of the profs of separating template and core is separate maintenance and(implicitly) ability to quickly add fixes/backports/shims to extension template, without waiting for new CKAN release. But, maybe, it would be nice to have(in addition to the explicit `cookiecutter URL` example) built-in command for template generation? It's unlikely that devs will remember template's URL and they will constantly pick this value from documentation(which may be irritating for those, who prefer to execute a sequence of simple commands). For example, if we would have `ckan generate extension` CLI command with hardcoded URL for template and ability to specify the name of the custom template and few additional options, then devs can just inspect CLI options(via `ckan --help`) and find out everything they need. It will create a basis for the generation of other kinds of artifacts, like `actions` and `plugins`.

f-osorio commented:
I think having an option to use a CLI command makes a lot of sense. There's some limitation in `cookiecutter` that introduce some awkwardness for the user that this could smooth over. I can work on that, but should it be in this pull request or a new one?

wardi commented:
The trouble with using a separate repo for the template is that anything other than ckan/ckan gets much less attention and testing from the core team. It's easy for a change in core to lead to a change in the best practices for new extensions, so I think it's better for the extension template to stay in core so that the changes can be made at the same time on the same branch.

If someone would like to maintain a separate template for extensions that is designed to work across a wide range of ckan versions (e.g. by importing ckan/ckantoolkit and only using "safe" ways of accessing core features) I fully support that effort, but that's a different problem than the one the default extension templates should solve.

amercader commented:
I agree with @wardi. @f-osorio Here's my suggested plan for this:

1. Include your template in this PR. Current template is in `ckan/pastertemplates`. Perhaps is best to put the new one in `contrib/cookiecutter/ckan_extension`
2. Add cookiecutter to `dev-requirements.txt`
3. Update the docs if needed

This should be the MVP. If you want to build a dedicated command to make it easier for users that's great but not a blocker for merging. It can go in this PR or a separate one. It should be a ["new-style"](https://github.com/ckan/ckan/issues/4639) command, eg `ckan extension create`. There are details on that issue about how to set them up and plenty of already done examples in `ckan/cli`

amercader commented:
Forgot to say thanks a lot for taking this on @f-osorio!

davidread commented:
I tried it out and it works a treat, @f-osorio :)

A couple of v. minor things:
* Spelling of 'seperate' -> 'separate' https://github.com/f-osorio/cookiecutter-ckan-extension/blob/master/cookiecutter.json#L3
* Compatibility is python 2.7 only, for the time being https://github.com/f-osorio/cookiecutter-ckan-extension/blob/master/%7B%7Bcookiecutter.project%7D%7D/setup.py#L46

davidread commented on doc/extensions/tutorial.rst:
```diff
@@ -65,18 +71,20 @@ contains the CKAN git repo. Let's use the parent directory instead:
+   extension. CKAN extension names *have* to begin with ``ckanext-``. This
+   tutorial uses the project name ``ckanext-iauthfunctions``.
+   The prompts for ``project_shortname`` and ``plugin_class_name`` can be
+   ignored. They will be filled in automatically.
```
The things mentioned in this note seem relatively clear to me during use. Does including the note just lengthen the docs unnecessarily?
```
$ cookiecutter https://github.com/f-osorio/cookiecutter-ckan-extension.git
project [Project's name (must begin with 'ckanext-')]: ckanext-hierarchy
keywords [A space seperated list of keywords]: tag1 tag2
description []: This is a great project!
author []: David
author_email []: david.read@hackneyworkshop.com
github_user_name [Github user or organization name]: davidread
project_shortname [hierarchy]:
plugin_class_name [HierarchyPlugin]:
```

f-osorio replied:
If it's clear then I agree that it's unnecessary. I wasn't sure how clear it is partly because it seems to me to be a little unintuitive to get a prompt for something that should be ignored. But there's no way around that.

`Cookiecutter` doesn't currently support having templates in subdirectories of repos. But there is a [PR for it and it's mostly through review](https://github.com/cookiecutter/cookiecutter/pull/1224).

wardi replied:
If cookiecutter doesn't support using a local path for a template then I'm fine with using another repo under the ckan org (kind of annoying for users to need internet access just to create an extension, but not worth maintaining a fork of cookiecutter)

f-osorio commented:
My mistake. I was thinking about someone using `cookiecutter` to get the template through a URL, but since the template should already be local it is possible to use a local path. I've updated the documentation to refer to this. 

There's also a CLI command in place. It's based on @smotornyuk suggestion, `ckan generate extension`, so that it can be expanded to generate other things should the need arise in the future.

The documentation right now goes through the process of using `cookiecutter <path/to/template>` but mentions the CLI. I wasn't sure how best to handle having two methods, and I couldn't find a place in the documentation that covers CLI.

There are some failures in the CI checks due to the syntax used by `cookiecutter`. I don't think there is anything that can be done about them within `cookiecutter`. The syntax is essential to how it works.

smotornyuk commented:
Perfect. As for thests, you can add template directory to whitelist - https://github.com/ckan/ckan/blob/dfad5d98a644a76939a57872073ef27dc7c68d86/ckan/tests/test_coding_standards.py#L32 . This should fix two of three failures. And third error should be fixed by listing problematic files here -https://github.com/ckan/ckan/blob/dfad5d98a644a76939a57872073ef27dc7c68d86/ckan/tests/legacy/test_coding_standards.py#L303

f-osorio commented:
I've updated the documentation to include using both `cookiecutter ...` and the CLI `ckan generate extension`. 

I added an exception to `.travis.yml` to ignore the directory `contrib/cookiecutter`. It seemed to be the only way to get the flake8 check to ignore the kind of weirdnesses that cookiecutter needs.

smotornyuk commented:
Cool! I don't see a problem with just ignoring this directory - this PR already does much more, that was supposed initially
